iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+template <class C>
+int
+RefCountCacheSync<C>::write_to_disk(char *i, int size)
+{
+  int written = 0;
+  while (written < size) {
+    int ret = this->socket_manager.write(this->fd, i + written, size - written);
+    if (ret <= 0) {
+      return -1;
+    } else {
+      written += ret;
+    }
+  }
+  return 0;
+}
+
+template <class C>
+RefCountCacheSync<C>::RefCountCacheSync(Continuation *acont, C *cc, int frequency, std::string dirname, std::string filename)
+  : Continuation(NULL),
+    partition(0),
+    cc(cc),
+    cont(acont),
+    fd(0),
+    dirname(dirname),
+    filename(filename),
+    time_per_partition(HRTIME_SECONDS(frequency) / cc->partition_count()),
+    start(Thread::get_hrtime()),
+    total_items(0),
+    total_size(0),
+    rsb(cc->get_rsb())
+
+{
+  eventProcessor.schedule_imm(this, ET_TASK);
+  mutex              = cc->get_partition(partition).lock.get();
+  this->tmp_filename = this->filename + ".syncing"; // TODO tmp file extension configurable?
+
+  SET_HANDLER((CacheSyncHandler)&RefCountCacheSync::initialize_storage);
+}
+
+// RefCountCache is a ref-counted key->value map to store classes that inherit from RefCountObj.
+// Once an item is `put` into the cache, the cache will maintain a Ptr<> to that object until erase
+// or clear is called-- which will remove the cache's Ptr<> to the object.
+// This cache may be Persisted (RefCountCacheSync) as well as loaded from disk (LoadRefCountCacheFromPath).
+// This class will optionally emit metrics at the given `metrics_prefix`.
+// Note: although this cache does allow you to set expiry times this cache does not actively GC itself-- meaning
+// it will only remove expired items once the space is required. So to ensure that the cache is bounded either a
+// size or an item limit must be set-- otherwise the cache will not GC. Also note, that if keys collide the previous
+// entry for a given key will be removed, so this "leak" concern is assuming you don't have sufficient space to store
+// an item for each possible key
+template <class C> class RefCountCache
+{
+public:
+  // Constructor
+  RefCountCache(unsigned int num_partitions, int size = -1, int items = -1, VersionNumber object_version = VersionNumber(),
+                std::string metrics_prefix = "");
+  // Destructor
+  ~RefCountCache();
+
+  // User interface to the cache
+  Ptr<C> get(uint64_t key);
+  void put(uint64_t key, C *item, int size = 0, int expiry_time = -1);
+  void erase(uint64_t key);
+  void clear();
+
+  // Some methods to get some internal state
+  int partition_for_key(uint64_t key);
+  ProxyMutex *lock_for_key(uint64_t key);
+  const size_t partition_count();
+  RefCountCachePartition<C> &get_partition(int pnum);
+  const size_t count();
+  RefCountCacheHeader &get_header();
+  RecRawStatBlock *get_rsb();
+
+private:
+  int max_size;  // Total size
+  int max_items; // Total number of items allowed
+
+  unsigned int num_partitions;
+
+  std::string filepath;
+  std::string dirname;
```

Comment:
You should not need to store this again here. ``HostDBSync`` has a copy and ``RefCountCacheSync`` has a copy, we should be able to without this one.